8261671: X86 I2L conversion can be skipped for certain masked positive values

mgkwill opened the pull request:
8261671: X86 I2L conversion can be skipped for certain masked positive values

For the following expression:
 (long)(value & mask)
Where value is of int type and mask is constant (power of two – 1), we can directly generate bzhi instruction to zero the upper bits instead of doing an andl, followed by movslq

Before:
Benchmark Mode Cnt Score Error Units
SkipIntToLongCast.skipMaskedSmallPositiveCast avgt 15 10.679 ± 1.496 ns/op


After:
Benchmark Mode Cnt Score Error Units
SkipIntToLongCast.skipMaskedSmallPositiveCast avgt 15 7.870 ± 0.067 ns/op

Signed-off-by: Marcus G K Williams <marcus.williams@intel.com>

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261671](https://bugs.openjdk.java.net/browse/JDK-8261671): X86 I2L conversion can be skipped for certain masked positive values


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to ca1d628e00231238f83afbdd46336852155b2bf4
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to c3bcaf444858125ab61e1b0900295d5ddcf30613
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 705c600118e10495a53ad2ab8b5e428b583cdd53


### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2590/head:pull/2590`
`$ git checkout pull/2590`

To update a local copy of the PR:
`$ git checkout pull/2590`
`$ git pull https://git.openjdk.java.net/jdk pull/2590/head`

mgkwill commented:
/label hotspot-compiler

fweimer-rh commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -9153,6 +9166,22 @@ instruct andI2L_rReg_imm65535(rRegL dst, rRegI src, immI_65535 mask)
+// Can skip int2long conversions after AND with small bitmask
+instruct convI2LAndI_reg_immIbitmask(rRegL dst, rRegI src, immI_bitmask mask, rRegI tmp, rFlagsReg cr)
+%{
+  predicate(VM_Version::supports_bmi2());
```
Agner's optimization guide says that BZHI uses microcode on Zen 2 and earlier, so perhaps the predicate should reflect that?

mgkwill replied:
Added `predicate(VM_Version::supports_bmi2() && VM_Version::is_intel());`

fweimer-rh replied:
I'm so sorry, I confused this with `PDEP` and `PEXT`. `BZHI` is actually fine on Zen. So the original code is correct.

mgkwill replied:
No worries @fweimer. I've reverted to the original. I looked at Agner's Optimization Guide but I couldn't find a place that said Zen 2 and earlier used microcode for BZHI. I guess we know why now :smiley:. 

Happy to hear any other review comments.

vnkozlov reviewed (changes requested):
Please, add a test which verifies result of generated code for different ranges of mask (especially corner cases). See `compiler//codegen/BMI1.java`
There are also example of verification of generated assembler in `compiler//intrinsics/bmi/`

mgkwill commented:
> Please, add a test which verifies result of generated code for different ranges of mask (especially corner cases). See `compiler//codegen/BMI1.java`
> There are also example of verification of generated assembler in `compiler//intrinsics/bmi/`

I'm taking a look at a test to verify generated code like the BMI1 test you suggest above.

There is a current test that verifies operation indirectly (both long to int and int to long): test/hotspot/jtreg/compiler/c2/TestSkipLongToIntCast.java

Do we need both? A number of similar patches have been merged without, https://hg.openjdk.java.net/jdk/jdk/rev/d0f55423e913 for example.

vnkozlov commented:
> > Please, add a test which verifies result of generated code for different ranges of mask (especially corner cases). See `compiler//codegen/BMI1.java`
> > There are also example of verification of generated assembler in `compiler//intrinsics/bmi/`
> 
> I'm taking a look at a test to verify generated code like the BMI1 test you suggest above.
> 
> There is a current test that verifies operation indirectly (both long to int and int to long): test/hotspot/jtreg/compiler/c2/TestSkipLongToIntCast.java

It does not cover your case which have `AND` operation.
But I am fine if you add your case to it (or to other test, like BMI1.java) instead of creating new test file.

> 
> Do we need both? A number of similar patches have been merged without, https://hg.openjdk.java.net/jdk/jdk/rev/d0f55423e913 for example.

Missing tests in previous changes does not mean we should never create one.

mgkwill commented:
> > > Please, add a test which verifies result of generated code for different ranges of mask (especially corner cases). See `compiler//codegen/BMI1.java`
> > > There are also example of verification of generated assembler in `compiler//intrinsics/bmi/`
> > 
> > 
> > I'm taking a look at a test to verify generated code like the BMI1 test you suggest above.
> > There is a current test that verifies operation indirectly (both long to int and int to long): test/hotspot/jtreg/compiler/c2/TestSkipLongToIntCast.java
> 
> It does not cover your case which have `AND` operation.
> But I am fine if you add your case to it (or to other test, like BMI1.java) instead of creating new test file.
> 
> > Do we need both? A number of similar patches have been merged without, https://hg.openjdk.java.net/jdk/jdk/rev/d0f55423e913 for example.
> 
> Missing tests in previous changes does not mean we should never create one.

Thanks @vnkozlov. 

In the process of testing both a BMI2.java and compiler/intrinsics/bmi/verifycode/BzhiTest.java addition I have written to test this change.

mgkwill commented:
Hi @vnkozlov. I've added a BMI2.java and a compiler/intrinsics/bmi/verifycode/BzhiTestI2L.java test as you suggested. I'd appreciate any further review or comments.

vnkozlov commented:
Thank you! I will run our testing and let you know results.

vnkozlov commented on test/hotspot/jtreg/compiler/codegen/BMI2.java:
```diff
@@ -0,0 +1,442 @@
+        { // match(Set dst (ConvI2L (AndI src1 mask)))
+            long z = BMITests.bzhiI2L(ix, i);
+
+            for (int i2 = 0; i2 < ITERATIONS; i2++) {
```
I know that you follow BMI.java code example but we need to update these tests. Since you are writing new test you may do it right. See discussion for https://github.com/openjdk/jdk/pull/2811

And, please, don't use 1000,000 iterations. C2 compilation is triggered with 10,000. So 50,000 would be more than enough.

mgkwill replied:
Thanks. I will have an update that lowers iterations. I've read through #2811 and will see what I can apply to BMI2.java tests.

vnkozlov commented:
compiler/intrinsics/bmi/TestBzhiI2L.java test failed:
```
 stdout: [For random generator using seed: -6986231732124296613
To re-run test with same seed value please add "-Djdk.test.lib.random.seed=-6986231732124296613" to command line.
];
 stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
	at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)
	at compiler.intrinsics.bmi.BMITestRunner$Executor.runUnaryLongRegTest(BMITestRunner.java:280)
	at compiler.intrinsics.bmi.BMITestRunner$Executor.runTests(BMITestRunner.java:219)
	at compiler.intrinsics.bmi.BMITestRunner$Executor.main(BMITestRunner.java:190)
]
 exitValue = 1

java.lang.RuntimeException: Expected to get exit value of [0]

	at jdk.test.lib.process.OutputAnalyzer.shouldHaveExitValue(OutputAnalyzer.java:474)
	at compiler.intrinsics.bmi.BMITestRunner.runTest(BMITestRunner.java:151)
	at compiler.intrinsics.bmi.BMITestRunner.runTests(BMITestRunner.java:87)
	at compiler.intrinsics.bmi.TestBzhiI2L.main(TestBzhiI2L.java:49)
```

mgkwill commented:
> compiler/intrinsics/bmi/TestBzhiI2L.java test failed:
> 
> ```
>  stdout: [For random generator using seed: -6986231732124296613
> To re-run test with same seed value please add "-Djdk.test.lib.random.seed=-6986231732124296613" to command line.
> ];
>  stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
> 	at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runUnaryLongRegTest(BMITestRunner.java:280)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runTests(BMITestRunner.java:219)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.main(BMITestRunner.java:190)
> ]
>  exitValue = 1
> 
> java.lang.RuntimeException: Expected to get exit value of [0]
> 
> 	at jdk.test.lib.process.OutputAnalyzer.shouldHaveExitValue(OutputAnalyzer.java:474)
> 	at compiler.intrinsics.bmi.BMITestRunner.runTest(BMITestRunner.java:151)
> 	at compiler.intrinsics.bmi.BMITestRunner.runTests(BMITestRunner.java:87)
> 	at compiler.intrinsics.bmi.TestBzhiI2L.main(TestBzhiI2L.java:49)
> ```

Thanks @vnkozlov. Interesting. I'll take a look using "-Djdk.test.lib.random.seed=-6986231732124296613"

vnkozlov commented:
Note, the test only failed on machines which do not support BMI2: AMD EPYC and Aarch64.

vnkozlov commented:
Except one failure on machine with i7: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 FMA CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC MOVBE POPCNT AES PCID XSAVE OSXSAVE SEGLIM64 TSCTMR AVX1.0 RDRAND F16C RDWRFSGS TSC THREAD OFFSET SGX BMI1 HLE AVX2 SMEP BMI2 ERMS INVPCID RTM FPU CSDS MPX RDSEED ADX SMAP CLFSOPT IPT SGXLC MDCLEAR TSXFA IBRS STIBP L1DF SSBD SYSCALL XD 1GBPAGE EM64T LAHF LZCNT PREFETCHW RDTSCP TSCI

So I am not sure about cause of failure.

mgkwill commented:
> Note, the test only failed on machines which do not support BMI2: AMD EPYC and Aarch64.

AH! Thanks. I couldn't reproduce here. I was writing a long response about that.

I'll take a look at the predicate and the test to see if I can exclude systems that cannot run BMI2. 

The predicate already had a check for BMI2: 
`predicate(VM_Version::supports_bmi2());`

and the bzhiq assembler method had a check:
`assert(VM_Version::supports_bmi2(), "bit manipulation instructions not supported");`



Here is a warning in Test, though this doesn't preclude running on non-BMI2 supporting processors.
```
if (!CPUInfo.hasFeature("bmi2")) {
            System.out.println("INFO: CPU does not support bmi2 feature.");
}
```

That i7 failure is concerning however. I'll look into that also. Thanks @vnkozlov

mgkwill commented:
> compiler/intrinsics/bmi/TestBzhiI2L.java test failed:
> 
> ```
>  stdout: [For random generator using seed: -6986231732124296613
> To re-run test with same seed value please add "-Djdk.test.lib.random.seed=-6986231732124296613" to command line.
> ];
>  stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
> 	at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runUnaryLongRegTest(BMITestRunner.java:280)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runTests(BMITestRunner.java:219)
> 	at compiler.intrinsics.bmi.BMITestRunner$Executor.main(BMITestRunner.java:190)
> ]
>  exitValue = 1
> 
> java.lang.RuntimeException: Expected to get exit value of [0]
> 
> 	at jdk.test.lib.process.OutputAnalyzer.shouldHaveExitValue(OutputAnalyzer.java:474)
> 	at compiler.intrinsics.bmi.BMITestRunner.runTest(BMITestRunner.java:151)
> 	at compiler.intrinsics.bmi.BMITestRunner.runTests(BMITestRunner.java:87)
> 	at compiler.intrinsics.bmi.TestBzhiI2L.main(TestBzhiI2L.java:49)
> ```

Hi @vnkozlov - 

- I've updated BMI2.java with a reduced iterations of 30K and I tried to incorporate your feedback on code structure from #2811. My previous version was an attempt to match the code style and structure used in existing code. We can agree that those tests are less than optimal. Please point out further specifics on anything that could be improved.

- On TestBzhiI2L.java I did some digging. It appears there were some implementation issues with the underlying BmiIntrinsicBase.java and the BmiTestCase_x64 class. I believe I've fixed those issues.

However in my estimation the first version of code should not have hit Expr.java:97 longExpr below:
>  stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
> 	at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)

It should have hit intToLongExpr. This prompted me to see that my tests did not include a couple of key parts to the BMI Test running setup, including a BMIUnaryIntToLongExpr in Expr.java that is used instead of BMIUnaryLongExpr. The test seemed to be working without these but I'm guessing there are corner cases that this would cause problems, maybe explaining what was hit above. Though I don't have a good explanation why yet.

I've made the fixes I can see were needed. I've sourced an I7 to see if I can reproduce the failure. In the meantime, assuming pre-submit tests pass, can you run your tests again?

Thanks,
Marcus

mgkwill commented:
> > compiler/intrinsics/bmi/TestBzhiI2L.java test failed:
> > ```
> >  stdout: [For random generator using seed: -6986231732124296613
> > To re-run test with same seed value please add "-Djdk.test.lib.random.seed=-6986231732124296613" to command line.
> > ];
> >  stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
> > 	at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)
> > 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runUnaryLongRegTest(BMITestRunner.java:280)
> > 	at compiler.intrinsics.bmi.BMITestRunner$Executor.runTests(BMITestRunner.java:219)
> > 	at compiler.intrinsics.bmi.BMITestRunner$Executor.main(BMITestRunner.java:190)
> > ]
> >  exitValue = 1
> > 
> > java.lang.RuntimeException: Expected to get exit value of [0]
> > 
> > 	at jdk.test.lib.process.OutputAnalyzer.shouldHaveExitValue(OutputAnalyzer.java:474)
> > 	at compiler.intrinsics.bmi.BMITestRunner.runTest(BMITestRunner.java:151)
> > 	at compiler.intrinsics.bmi.BMITestRunner.runTests(BMITestRunner.java:87)
> > 	at compiler.intrinsics.bmi.TestBzhiI2L.main(TestBzhiI2L.java:49)
> > ```
> 
> Hi @vnkozlov -
> 
> * I've updated BMI2.java with a reduced iterations of 30K and I tried to incorporate your feedback on code structure from #2811. My previous version was an attempt to match the code style and structure used in existing code. We can agree that those tests are less than optimal. Please point out further specifics on anything that could be improved.
> * On TestBzhiI2L.java I did some digging. It appears there were some implementation issues with the underlying BmiIntrinsicBase.java and the BmiTestCase_x64 class. I believe I've fixed those issues.
> 
> However in my estimation the first version of code should not have hit Expr.java:97 longExpr below:
> 
> > stderr: [Exception in thread "main" java.lang.UnsupportedOperationException
> > at compiler.intrinsics.bmi.Expr.longExpr(Expr.java:97)
> 
> It should have hit intToLongExpr. This prompted me to see that my tests did not include a couple of key parts to the BMI Test running setup, including a BMIUnaryIntToLongExpr in Expr.java that is used instead of BMIUnaryLongExpr. The test seemed to be working without these but I'm guessing there are corner cases that this would cause problems, maybe explaining what was hit above. Though I don't have a good explanation why yet.
> 
> I've made the fixes I can see were needed. I've sourced an I7 to see if I can reproduce the failure. In the meantime, assuming pre-submit tests pass, can you run your tests again?
> 
> Thanks,
> Marcus

P.S. I added some specific logic to skip TestBzhiI2L.java tests if BMI2 is not available and hopefully fixed the logic that should have stopped the tests from running before if cpuFeature was not present. Hopefully this solves the issue involving AMD EPYC and Aarch64.

vnkozlov commented:
Good.
Unfortunately we need to wait next week for testing results because we are updating our testing system.

mgkwill commented:
> Good.
> Unfortunately we need to wait next week for testing results because we are updating our testing system.

Thanks. I'll check in with you next week.

vnkozlov commented:
Testing was restored and I ran tier104 as before - no failed tests.
But 2 new tests filed failed copyright headers varification:
```
BAD COPYRIGHT LINE: /workspace/open/test/hotspot/jtreg/compiler/intrinsics/bmi/verifycode/BzhiTestI2L.java
BAD COPYRIGHT LINE: /workspace/open/test/hotspot/jtreg/compiler/intrinsics/bmi/TestBzhiI2L.java
```

vnkozlov commented:
It is missing `,` after year:
```
Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
```

mgkwill commented:
> It is missing `,` after year:
> 
> ```
> Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
> ```

@vnkozlov Thanks for all your help and review. I'll push an update tomorrow with fixed copyright headers.

mgkwill commented:
> It is missing `,` after year:
> 
> ```
> Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
> ```

@vnkozlov Done.

vnkozlov reviewed (approved):
Good.

You need second review before push.

mgkwill commented:
> Good.
> 
> You need second review before push.

Thanks @vnkozlov. I will wait for another approval.

neliasso commented on test/hotspot/jtreg/compiler/codegen/BMI2.java:
```diff
@@ -0,0 +1,449 @@
+ * @summary Support BMI2 instructions on x86/x64
+ *
+ * @run main/othervm -Xbatch -XX:-TieredCompilation -XX:+UnlockDiagnosticVMOptions -XX:+AbortVMOnCompilationFailure
+ *      -XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*
```
It might be better to use "-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" Then the main methods can still be compiled. As long as BMI2$BMITests method isn't inlining other BMI2$BMITests methods this is will just work just fine.

mgkwill replied:
Sounds reasonable.

mgkwill replied:
I tried to use "-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" however I'm unable to directly verify the C2 compilation of bzhiI2L() method and therefor bzhiq instructions that should be generated. I see the testZeroBzhiI2L() testOneBzhiI2L() methods get compiled and they both make a static call to  bzhiI2L() but it doesn't appear to be C2 compiled.

I'm using the following command: 'make test TEST="jtreg:test/hotspot/jtreg/compiler/codegen/BMI2.java" JTREG="VM_OPTIONS=-XX:+UnlockDiagnosticVMOptions -XX:-TieredCompilation -XX:+PrintAssembly -Xbatch -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::* -XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" CONF=linux-x86_64-server-slowdebug'

But I've tried many variations. If I use "-XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*" I can manually verify that bzhiq is generated by C2 compiler.

I'd prefer to leave this as is unless I'm missing something.

vnkozlov replied:
Nils's suggestion is reasonable. With "compileonly,compiler.codegen.BMI2$BMITests::*" the test will be slow because BMI2::test*BzhiI2L() methods are not compiled.
 Use -XX:+PrintCompilation to see what is going on with compilations. I think you will see several recompilations until all switch cases are executed. That is why I want loops in test*BzhiI2L() be inverted (ITERATIONS as external loop) - you will get final code version much sooner. You can preserve golden values (first run) in array to compare later.
BTW, you don't need 2 methods - pass `ix` as parameter.

```
public static void testBzhiI2L(int ix) {
       long[] goldv = new long[16];
       for (int i = 0; i <= 15; i++) {
             goldv[i] = BMITests.bzhiI2L(ix, i);
       }
       for (int i2 = 0; i2 < ITERATIONS; i2++) {
            for (int i = 0; i <= 15; i++) {
                 long v = BMITests.bzhiI2L(ix, i);
                 if (v != goldv[i]) {
                         throw new Error(returnBzhiI2LErrMessage (gold[i], v));
                 }
            }
      }
}
```

mgkwill replied:
Thanks @vnkozlov and @neliasso. I've made the changes you suggest.

I'm still unsure about `-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*` over `-XX:CompileCommand=compileonly,compiler.codegen.BMI2$BMITests::*` but I will trust in your knowledge and experience in this area. It doesn't materially change the test just how it's run by default and the test should C2 compile and compare 
`bzhiq`
version to 
```
andI
movslq 
```

Even  with -XX:+PrintCompilation I'm not able to see this happening in test .jtr using  `-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*` and printing assembly, inlining, compilation but perhaps I am reading the logs incorrectly.

neliasso commented on test/hotspot/jtreg/compiler/codegen/BMI2.java:
```diff
@@ -0,0 +1,449 @@
+
+            z = BMITests.bzhiI2L(ix, i);
+
+            for (int i2 = 0; i2 < ITERATIONS; i2++) {
```
Why are these manually unrolled? They are the same but with different index?

mgkwill replied:
I wasn't able to get the above to C2 compile separately. I wrote essentially the same but could not verify the compilation. I then unrolled from the for loop and could verify the bzhi instructions were generated. Repetition felt like a reasonable compromise for completeness. I'll take another look and see if I can verify in a for loop. Perhaps one of the options I was using prevented it from working.

mgkwill replied:
Rolling this into for loops appears to work. I'm guessing something else was at issue before when I couldn't verify bzhiq instructions in C2 compilation of bzhiI2L(). Pushed a patch with the change.

mgkwill commented:
Thanks for the review @neliasso. I tried both of your suggestions. 
- Rolling the test operations into a 'for' loop worked, it appears my earlier troubles w/ C2 compilation were due to another factor. 
- Using "-XX:CompileCommand=dontinline,compiler.codegen.BMI2$BMITests::*" did not allow me to manually verify the assembly code was appearing and that the test was valid. Perhaps I'm missing something. See my comment for more detail.

Thanks again for your review. Please let me know what you think and if you have other suggestions or comments.

vnkozlov commented:
I think you become "victim" of JTREG's feature which removes middle of long output:

```
...
Output overflow:
JT Harness has limited the test output to the text
at the beginning and the end, so that you can see how the
test began, and how it completed.
```
You need to remove `-XX:+PrintAssembly` option which prints code for ALL compiled method and overflow jtreg output.
You only need `-XX:CompileCommand=print,class::method` which prints assembler for specified method.

An other issue is that `make test` "eats" `$` in class name. I ran command you used:
`make test TEST="jtreg:test/hotspot/jtreg/compiler/codegen/BMI2.java" JTREG="VM_OPTIONS=-XX:+UnlockDiagnosticVMOptions -XX:-TieredCompilation -Xbatch -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::*"`

and got next message in output .jtr file:

```
----------System.out:(2/154)----------
CompileCommand: print compiler/codegen/BMI2MITests.* bool print = true
CompileCommand: dontinline compiler/codegen/BMI2$BMITests.* bool dontinline = true
```
`dontinline` line comes from test `@run` command and it is correct. But class name for `print` is wrong. I tried to use `\$` - did not help.

I ran the test with `jtreg`:
```
jtreg -testjdk:my_jdk -va -javaoptions:'-server -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::* -XX:-TieredCompilation -Xbatch' compiler/codegen/BMI2.java
```
And got nice code for 
```
04c     B4: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
04c     bzhiq RAX, RSI, #32767  # using R10 as TEMP, int & immI_bitmask -> long
057     jmp     B20

05c     B5: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
05c     bzhiq RAX, RSI, #16383  # using R10 as TEMP, int & immI_bitmask -> long
067     jmp     B20

06c     B6: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
06c     bzhiq RAX, RSI, #8191   # using R10 as TEMP, int & immI_bitmask -> long
077     jmp     B20

07c     B7: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
07c     bzhiq RAX, RSI, #4095   # using R10 as TEMP, int & immI_bitmask -> long
087     jmp     B20

08c     B8: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
08c     bzhiq RAX, RSI, #2047   # using R10 as TEMP, int & immI_bitmask -> long
097     jmp,s   B20

099     B9: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
099     bzhiq RAX, RSI, #1023   # using R10 as TEMP, int & immI_bitmask -> long
0a4     jmp,s   B20

0a6     B10: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0a6     bzhiq RAX, RSI, #511    # using R10 as TEMP, int & immI_bitmask -> long
0b1     jmp,s   B20

0b3     B11: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0b3     movzbl  RAX, RSI        # int & 0xFF -> long
0b7     jmp,s   B20

0b9     B12: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0b9     bzhiq RAX, RSI, #127    # using R10 as TEMP, int & immI_bitmask -> long
0c4     jmp,s   B20

0c6     B13: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0c6     bzhiq RAX, RSI, #63     # using R10 as TEMP, int & immI_bitmask -> long
0d1     jmp,s   B20

0d3     B14: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0d3     bzhiq RAX, RSI, #31     # using R10 as TEMP, int & immI_bitmask -> long
0de     jmp,s   B20

0e0     B15: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0e0     bzhiq RAX, RSI, #15     # using R10 as TEMP, int & immI_bitmask -> long
0eb     jmp,s   B20

0ed     B16: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0ed     bzhiq RAX, RSI, #7      # using R10 as TEMP, int & immI_bitmask -> long
0f8     jmp,s   B20

0fa     B17: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
0fa     bzhiq RAX, RSI, #3      # using R10 as TEMP, int & immI_bitmask -> long
105     jmp,s   B20

107     B18: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
107     bzhiq RAX, RSI, #1      # using R10 as TEMP, int & immI_bitmask -> long
112     jmp,s   B20

114     B19: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
114     movzwl  RAX, RSI        # int & 0xFFFF -> long

117     B20: #  out( N1 ) <- in( B19 B4 B5 B6 B7 B8 B9 B10 B11 B12 B13 B14 B15 B16 B17 B18 )  Freq: 0.999999
117     addq    rsp, 32 # Destroy frame
        popq    rbp
        cmpq     rsp, poll_offset[r15_thread]
        ja       #safepoint_stub        # Safepoint: poll for GC

129     ret
```

You patch and test are good!

mgkwill commented:
> I think you become "victim" of JTREG's feature which removes middle of long output:
> 
> ```
> ...
> Output overflow:
> JT Harness has limited the test output to the text
> at the beginning and the end, so that you can see how the
> test began, and how it completed.
> ```
> 
> You need to remove `-XX:+PrintAssembly` option which prints code for ALL compiled method and overflow jtreg output.
> You only need `-XX:CompileCommand=print,class::method` which prints assembler for specified method.
> 
> An other issue is that `make test` "eats" `$` in class name. I ran command you used:
> `make test TEST="jtreg:test/hotspot/jtreg/compiler/codegen/BMI2.java" JTREG="VM_OPTIONS=-XX:+UnlockDiagnosticVMOptions -XX:-TieredCompilation -Xbatch -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::*"`
> 
> and got next message in output .jtr file:
> 
> ```
> ----------System.out:(2/154)----------
> CompileCommand: print compiler/codegen/BMI2MITests.* bool print = true
> CompileCommand: dontinline compiler/codegen/BMI2$BMITests.* bool dontinline = true
> ```
> 
> `dontinline` line comes from test `@run` command and it is correct. But class name for `print` is wrong. I tried to use `\$` - did not help.
> 
> I ran the test with `jtreg`:
> 
> ```
> jtreg -testjdk:my_jdk -va -javaoptions:'-server -XX:CompileCommand=print,compiler.codegen.BMI2$BMITests::* -XX:-TieredCompilation -Xbatch' compiler/codegen/BMI2.java
> ```
> 
> And got nice code for
> 
> ```
> 04c     B4: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 04c     bzhiq RAX, RSI, #32767  # using R10 as TEMP, int & immI_bitmask -> long
> 057     jmp     B20
> 
> 05c     B5: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 05c     bzhiq RAX, RSI, #16383  # using R10 as TEMP, int & immI_bitmask -> long
> 067     jmp     B20
> 
> 06c     B6: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 06c     bzhiq RAX, RSI, #8191   # using R10 as TEMP, int & immI_bitmask -> long
> 077     jmp     B20
> 
> 07c     B7: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 07c     bzhiq RAX, RSI, #4095   # using R10 as TEMP, int & immI_bitmask -> long
> 087     jmp     B20
> 
> 08c     B8: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 08c     bzhiq RAX, RSI, #2047   # using R10 as TEMP, int & immI_bitmask -> long
> 097     jmp,s   B20
> 
> 099     B9: #   out( B20 ) <- in( B2 )  Freq: 0.0624999
> 099     bzhiq RAX, RSI, #1023   # using R10 as TEMP, int & immI_bitmask -> long
> 0a4     jmp,s   B20
> 
> 0a6     B10: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0a6     bzhiq RAX, RSI, #511    # using R10 as TEMP, int & immI_bitmask -> long
> 0b1     jmp,s   B20
> 
> 0b3     B11: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0b3     movzbl  RAX, RSI        # int & 0xFF -> long
> 0b7     jmp,s   B20
> 
> 0b9     B12: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0b9     bzhiq RAX, RSI, #127    # using R10 as TEMP, int & immI_bitmask -> long
> 0c4     jmp,s   B20
> 
> 0c6     B13: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0c6     bzhiq RAX, RSI, #63     # using R10 as TEMP, int & immI_bitmask -> long
> 0d1     jmp,s   B20
> 
> 0d3     B14: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0d3     bzhiq RAX, RSI, #31     # using R10 as TEMP, int & immI_bitmask -> long
> 0de     jmp,s   B20
> 
> 0e0     B15: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0e0     bzhiq RAX, RSI, #15     # using R10 as TEMP, int & immI_bitmask -> long
> 0eb     jmp,s   B20
> 
> 0ed     B16: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0ed     bzhiq RAX, RSI, #7      # using R10 as TEMP, int & immI_bitmask -> long
> 0f8     jmp,s   B20
> 
> 0fa     B17: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 0fa     bzhiq RAX, RSI, #3      # using R10 as TEMP, int & immI_bitmask -> long
> 105     jmp,s   B20
> 
> 107     B18: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 107     bzhiq RAX, RSI, #1      # using R10 as TEMP, int & immI_bitmask -> long
> 112     jmp,s   B20
> 
> 114     B19: #  out( B20 ) <- in( B2 )  Freq: 0.0624999
> 114     movzwl  RAX, RSI        # int & 0xFFFF -> long
> 
> 117     B20: #  out( N1 ) <- in( B19 B4 B5 B6 B7 B8 B9 B10 B11 B12 B13 B14 B15 B16 B17 B18 )  Freq: 0.999999
> 117     addq    rsp, 32 # Destroy frame
>         popq    rbp
>         cmpq     rsp, poll_offset[r15_thread]
>         ja       #safepoint_stub        # Safepoint: poll for GC
> 
> 129     ret
> ```
> 
> You patch and test are good!

Thanks @vnkozlov Vladimir. That is really helpful! I'll keep in mind jtreg's output limiter and the 'make test' issue with class specifier in the future.

I will wait for another approval before integrating.

mgkwill commented:
Hello @neliasso & hotspot-compiler-dev, 

Can anyone else review this change and sign-off if you think it looks good?

Thanks,
Marcus

sviswa7 commented:
Hi @neliasso Marcus has incorporated your comments. Please review and approve if the changes look ok to you.

iwanowww commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -3176,6 +3176,19 @@ operand immL_32bits()
   interface(CONST_INTER);
 %}
 
+// Int Immediate: power of 2, positive
```
The comment is misleading: it's not a 2^n, but 2^n-1.

mgkwill replied:
👍 - Agreed, this is more descriptive.

iwanowww commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -3176,6 +3176,19 @@ operand immL_32bits()
+operand immI_bitmask()
+%{
+  predicate((n->get_int() != 0)
+            && ((n->get_int() & 0xc0000000) == 0)
```
Why do you use `& 0xc0000000` and not `& 0x80000000` (or even `n->get_int() > 0`)?

mgkwill replied:
Yes. I believe `n->get_int() > 0` would work and be simpler to read.

iwanowww commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -9153,6 +9166,22 @@ instruct andI2L_rReg_imm65535(rRegL dst, rRegI src, immI_65535 mask)
+  match(Set dst (ConvI2L (AndI src mask)));
+  format %{ "bzhiq $dst, $src, $mask \t# using $tmp as TEMP, int & immI_bitmask -> long" %}
+  ins_encode %{
+    assert(VM_Version::supports_bmi2(), "required");
```
The assert is redundant here. The predicate already guarantees that invariant.

mgkwill replied:
Agreed.

iwanowww commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -3176,6 +3176,19 @@ operand immL_32bits()
 %}
 
+// Int Immediate: power of 2, positive
+operand immI_bitmask()
```
`immI_bitmask` doesn't say much. What about `immI_Pow2M1`?

mgkwill replied:
👍 - Agreed, this is more descriptive.

neliasso reviewed (approved):
My comments has been resolved and the result looks good.

mgkwill commented:
Thanks for the review Vladimir ( @iwanowww ). I've incorporated your suggestions. Please let me know your thoughts or any more suggestions.

Thanks Nils ( @neliasso ) for your review and approval!

Thanks,
Marcus

iwanowww commented on src/hotspot/cpu/x86/x86_64.ad:
```diff
@@ -3176,6 +3176,19 @@ operand immL_32bits()
+// Int Immediate: 2^n-1, postive
+operand immI_Pow2M1()
+%{
+  predicate((n->get_int() != 0)
```
`n->get_int() != 0` is redundant now.

mgkwill replied:
Agreed. Update forthcoming.

iwanowww reviewed (approved):
Looks good.

mgkwill commented:
/integrate

sviswa7 commented:
/sponsor
